src/main/java/org/sonarlint/intellij/finding/persistence/FindingsManager.java

```diff
@@ -0,0 +1,247 @@
+
+import com.intellij.openapi.editor.RangeMarker;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.project.ProjectManager;
+import com.intellij.openapi.project.ProjectManagerListener;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.serviceContainer.NonInjectable;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.ThreadSafe;
+import org.jetbrains.annotations.NotNull;
+import org.sonarlint.intellij.common.ui.SonarLintConsole;
+import org.sonarlint.intellij.finding.LiveFinding;
+import org.sonarlint.intellij.finding.hotspot.LiveSecurityHotspot;
+import org.sonarlint.intellij.finding.tracking.Input;
+import org.sonarlint.intellij.finding.tracking.Trackable;
+import org.sonarlint.intellij.finding.tracking.Tracker;
+import org.sonarlint.intellij.finding.tracking.Tracking;
+import org.sonarlint.intellij.finding.issue.LiveIssue;
+import org.sonarlint.intellij.messages.IssueStoreListener;
+import org.sonarsource.sonarlint.core.commons.IssueSeverity;
+
+import static java.util.Collections.emptyList;
+
+/**
+ * Stores findings associated to a {@link PsiElement}, {@link RangeMarker} or  {@link PsiFile}.
+ * Findings are either issues or hotspots.
+ * Findings can then be displayed as annotations at any time.
+ */
+@ThreadSafe
+public class FindingsManager {
+  private final Project myProject;
+  private final LiveFindingCache<LiveIssue> liveIssueCache;
+  private final LiveFindingCache<LiveSecurityHotspot> liveSecurityHotspotCache;
+
+  public FindingsManager(Project project) {
+    this(project, new LiveFindingCache<>(project, new FindingPersistence<>(project, "issue")));
+  }
+
+  @NonInjectable
+  FindingsManager(Project project, LiveFindingCache<LiveIssue> liveIssueCache) {
+    myProject = project;
+    this.liveIssueCache = liveIssueCache;
+    this.liveSecurityHotspotCache = new LiveFindingCache<>(project, new FindingPersistence<>(project, "securityhotspot"));
+    project.getMessageBus().connect().subscribe(ProjectManager.TOPIC, new ProjectManagerListener() {
+      @Override
+      public void projectClosing(@NotNull Project project) {
+        try {
+          if (project == myProject) {
+            // Flush findings before project is closed, because we need to resolve module paths to compute the key
+            liveIssueCache.flushAll();
+            liveSecurityHotspotCache.flushAll();
+          }
+        } catch (Exception e) {
+          SonarLintConsole.get(myProject).error("Cannot flush issues", e);
+        }
+      }
+    });
+  }
+
+  public void clearAllIssuesForAllFiles() {
+    liveIssueCache.clear();
+    myProject.getMessageBus().syncPublisher(IssueStoreListener.SONARLINT_ISSUE_STORE_TOPIC).allChanged();
```

Comment:
What about renaming IssueStoreListener to FindingStoreListener?

Reply:
Yes. And I hope to kill this interface in a next refactoring :)